Update  wcs_maybe_prefix_key() and wcs_maybe_unprefix_key() to support an array of keys

james-allan opened the pull request:
## Description

> ✏️ **Note:** PR description updated to reflect recent changes.

This is a simple PR which updates the helper functions `wcs_maybe_prefix_key()` and `wcs_maybe_unprefix_key()` to prefix an array of keys or statuses for database storage. 

Example uses:

```php
wcs_maybe_prefix_key( 'active', 'wc-' ); // wc-active
wcs_maybe_prefix_key( [ 'active', 'processing' ], 'wc-' ); // [ 'wc-active', 'wc-processing' ] );
```

This feature of this function isn't used anywhere yet but it is useful for our report classes which need to prefix multiple statuses. See the PR here: https://github.com/woocommerce/woocommerce-subscriptions/pull/4837


## How to test this PR

1. Test as part of https://github.com/woocommerce/woocommerce-subscriptions/pull/4837.
2. General code review. 

## Product impact
<!-- What products will this PR ship in? -->

- [x] Added changelog entry (or does not apply)
- [ ] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [ ] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref
- [ ] <!-- 🚨 Deprecations 🚨 --> Added deprecated functions, hooks or classes to the [spreadsheet](https://docs.google.com/spreadsheets/d/1xw9xszcPMnWsp4C8OKZMsLzZob7tOmWT7qMqmEIq314/edit#gid=0)

james-allan committed:
Add helper function wcs_prefix_order_status()

james-allan committed:
Add optional flag to param description

james-allan committed:
Add unit test

james-allan committed:
Add changelog entry

barryhughes commented on includes/wcs-helper-functions.php:
```diff
@@ -349,3 +349,15 @@ function wcs_compare_order_billing_shipping_address( $order ) {
+ * @return string The order status prefixed with "wc-".
+ */
+function wcs_prefix_order_status( $order_status, $prefix = 'wc-' ) {
+	return wcs_maybe_prefix_key( $order_status, $prefix );
```
Definitely not a deal-breaker of any kind, I'm quite happy if we prefer to run with the current version, but I noticed that most of the existing usages of this new function (those added in https://github.com/woocommerce/woocommerce-subscriptions/pull/4837) look like this:

```php
$active_statuses = array_map( 'wcs_prefix_order_status', apply_filters( 'wcs_reports_active_statuses', [ 'active', 'pending-cancel' ] ) );
```

If that's likely to also be true in other future uses, perhaps we should make this function also handle an array of `$order_statuses` and not just a single status string?

To re-word a little, if every usage so far is wrapped in an `array_map()`, perhaps that's a sign we should handle arrays at this level and in doing so make the call-sites a little cleaner/a little more concise?

```php
// Possible reworking, though there are a few ways to do this ...
function wcs_prefix_order_status( $order_status, $prefix = 'wc-' ) {
	return is_array( $order_status )
		? array_map( __FUNCTION__, $order_status, array_fill( 0, count( $order_status ), $prefix ) )
		: wcs_maybe_prefix_key( $order_status, $prefix );
}
```

james-allan replied:
👍 Yeah so the reason I added this function was because I found I was doing this a lot:

```php
array_map( 'wcs_maybe_prefix_key', $order_statuses, array_fill( 0, count( $order_status ), $prefix ) )
```

With the introduction of this function, I removed 1 level of repetitiveness by no longer needing the `array_fill()`. But I didn't remove the 2nd case of repetitiveness of always calling this with array map. 

I like the idea of this new function being a more generic "prefix this array of strings".

barryhughes commented on includes/wcs-helper-functions.php:
```diff
@@ -349,3 +349,15 @@ function wcs_compare_order_billing_shipping_address( $order ) {
+ *
+ * @return string The order status prefixed with "wc-".
+ */
+function wcs_prefix_order_status( $order_status, $prefix = 'wc-' ) {
```
Even if the initial intended use case is dealing with order statuses, it feels like a pretty generic operation that could be applied in all sorts of use cases. Is there any merit in renaming to something like `wcs_prefix_string`?

james-allan replied:
We have a `wcs_maybe_prefix_key()` function for the more generic case. I've updated this PR to just make use of it and add support for array of keys/strings.

barryhughes reviewed (commented):
I realize you haven't requested a review on this one just yet, but leaving some notes relating to my review of https://github.com/woocommerce/woocommerce-subscriptions/pull/4837

james-allan committed:
Delete wcs_prefix_order_status and update wcs_maybe_prefix_key() instead

james-allan committed:
Update wcs_maybe_unprefix_key tests to include array cases

james-allan committed:
Update changelog entry

james-allan commented:
@barryhughes I've updated this PR based on your feedback here: https://github.com/Automattic/woocommerce-subscriptions-core/pull/824#discussion_r2012562354.

We already had a generic `wcs_maybe_prefix_key()` function for prefixing keys. Initially, I introduced a new function specifically for order statuses, but I've now refactored things to extend `wcs_maybe_prefix_key()` to support arrays of keys instead.

I think this results in a cleaner, more broadly useful approach that also addresses both of your suggestions. Let me know what you think. :)

barryhughes reviewed (approved):
That's awesome, thanks—works great:

<div align="center"><img width="600" alt="" src="https://github.com/user-attachments/assets/336f8d73-ed4d-4ff8-a8df-beab75c2a6c5"></div<

barryhughes committed:
Merge branch 'trunk' into add/wcs_prefix_order_status
